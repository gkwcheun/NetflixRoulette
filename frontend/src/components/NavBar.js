import React from "react";
import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useLocation } from "react-router";
import { Link } from "react-router-dom";
import { fetchMovies } from "../redux/actions";
import Dropdown from "react-bootstrap/Dropdown";
import PropTypes from "prop-types";

function NavBar(props) {
  const { fetchMovies } = props;
  const [genre, setGenre] = useState(null);
  const [sortBy, setSortBy] = useState(null);
  const [order, setOrder] = useState(null);
  const [navBarStatus, setNavBarStatus] = useState({
    all: true,
    documentary: false,
    comedy: false,
    horror: false,
    crime: false,
  });
  const [dropDownText, setDropDownText] = useState("RELEASE DATE");

  const query = new URLSearchParams(useLocation().search);

  const resetNavBar = {
    all: false,
    documentary: false,
    comedy: false,
    horror: false,
    crime: false,
  };

  useEffect(() => {
    fetchMovies(genre, sortBy, order, null, null);
  }, [genre, sortBy, order]);

  const onNavChange = (e) => {
    const navName = e.target.getAttribute("name");
    if (navName === "all") {
      setGenre(null);
    } else {
      setGenre(navName);
    }
    setNavBarStatus({
      ...resetNavBar,
      [navName]: true,
    });
  };

  useEffect(() => {
    // there are genres that aren't in the figma UI, so not on the list of genres
    // how to incorporate or just only append whatever genres in UI to obj?
    const searchGenre = query.get("genre") && query.get("genre").toLowerCase();
    let dropDown = "RELEASE DATE";
    const sortBy = query.get("sortBy") && query.get("sortBy").toLowerCase();
    const sortOrder =
      query.get("sortOrder") && query.get("sortOrder").toLowerCase();
    if (sortBy && sortOrder) {
      if (sortBy === "release_date") {
        dropDown = "RELEASE DATE";
      } else if (sortBy === "vote_average") {
        dropDown = "HIGHEST RATED";
      }
      if (sortOrder === "asc") {
        dropDown += " ASCENDING";
      } else if (sortOrder === "desc") {
        dropDown += " DESCENDING";
      }
    }
    if (searchGenre) {
      setNavBarStatus({
        ...resetNavBar,
        [searchGenre]: true,
      });
    }
    setDropDownText(dropDown);
  }, []);

  const onDropDownChange = (e) => {
    if (e.target.getAttribute("name") === "clear") {
      setSortBy(null);
      setOrder(null);
      setDropDownText("RELEASE DATE");
    } else {
      let [criteria, sortOrder] = e.target.getAttribute("name").split("-");
      setSortBy(criteria);
      setOrder(sortOrder);
      setDropDownText(e.target.innerHTML.toUpperCase());
    }
  };

  return (
    <>
      <div className="nav-bar-container">
        <nav className="nav-bar">
          <Link
            to="/search"
            className={`nav-link ${navBarStatus.all ? "selected-nav" : null}`}
            onClick={onNavChange}
            name="all"
          >
            ALL
          </Link>
          <Link
            to="/search?genre=documentary"
            className={`nav-link ${
              navBarStatus.documentary ? "selected-nav" : null
            }`}
            onClick={onNavChange}
            name="documentary"
          >
            DOCUMENTARY
          </Link>
          <Link
            to="/search?genre=comedy"
            className={`nav-link ${
              navBarStatus.comedy ? "selected-nav" : null
            }`}
            onClick={onNavChange}
            name="comedy"
          >
            COMEDY
          </Link>
          <Link
            to="/search?genre=horror"
            className={`nav-link ${
              navBarStatus.horror ? "selected-nav" : null
            }`}
            onClick={onNavChange}
            name="horror"
          >
            HORROR
          </Link>
          <Link
            to="/search?genre=crime"
            className={`nav-link ${navBarStatus.crime ? "selected-nav" : null}`}
            onClick={onNavChange}
            name="crime"
          >
            CRIME
          </Link>
        </nav>
        <div className="sort-nav">
          <Dropdown>
            <Dropdown.Toggle
              data-testid="dropdown-toggle"
              variant="dark"
              id="dropdown-button-dark-example2"
              className="sortby-toggle"
            >
              <span className="sort-by-label">SORT BY</span>
              {dropDownText}
            </Dropdown.Toggle>
            <Dropdown.Menu variant="dark">
              <Dropdown.Item onClick={onDropDownChange}>
                <Link
                  className="nav-link"
                  to="/search?sortBy=release_date&sortOrder=asc"
                  name="release_date-asc"
                >
                  Release Date Ascending
                </Link>
              </Dropdown.Item>
              <Dropdown.Item onClick={onDropDownChange}>
                <Link
                  className="nav-link"
                  to="/search?sortBy=release_date&sortOrder=desc"
                  name="release_date-desc"
                >
                  Release Date Descending
                </Link>
              </Dropdown.Item>
              <Dropdown.Item
                onClick={onDropDownChange}
                name="vote_average-desc"
              >
                <Link
                  className="nav-link"
                  to="/search?sortBy=vote_average&sortOrder=desc"
                  name="vote_average-desc"
                >
                  Highest Rated
                </Link>
              </Dropdown.Item>
              <Dropdown.Item onClick={onDropDownChange}>
                <Link className="nav-link" to="/search" name="clear">
                  Clear Sort
                </Link>
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </div>
      </div>
    </>
  );
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchMovies: (
      genre = null,
      criteria = null,
      order = null,
      query = null,
      category = null
    ) => {
      return dispatch(fetchMovies(genre, criteria, order, query, category));
    },
  };
};

NavBar.propTypes = {
  fetchMovies: PropTypes.func.isRequired,
};

export default connect(null, mapDispatchToProps)(NavBar);
