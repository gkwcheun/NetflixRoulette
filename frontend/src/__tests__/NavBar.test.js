import React from "react";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import NavBar from "../components/NavBar";
import { BrowserRouter as Router } from "react-router-dom";

const mockStore = configureStore([]);

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    search: "",
  }),
}));

test("placeholder", () => {});

describe("Nav Bar Testing", () => {
  let store;
  let componentSnapshot;
  let component;
  let initialState = {};
  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    componentSnapshot = renderer.create(
      <Provider store={store}>
        <Router>
          <NavBar />
        </Router>
      </Provider>
    );
    component = render(
      <Provider store={store}>
        <Router>
          <NavBar />
        </Router>
      </Provider>
    );
  });
  test("Nav Bar Snapshot", () => {
    expect(componentSnapshot.toJSON()).toMatchSnapshot();
  });
  test("Nav Links Render", () => {
    const { getByText } = component;
    getByText("ALL");
    getByText("DOCUMENTARY");
    getByText("HORROR");
    getByText("CRIME");
  });
  test("Nav Bar ALL link is selected on render by default", () => {
    const { getByText } = component;
    const allLink = getByText("ALL");
    expect(allLink.className).toBe("nav-link selected-nav");
  });
  test("Clicking NavBar links changes classname to selected", () => {
    const { getByText } = component;
    const allLink = getByText("ALL");
    const documentaryLink = getByText("DOCUMENTARY");
    const horrorLink = getByText("HORROR");
    const crimeLink = getByText("CRIME");
    // initially ALL is selected
    expect(allLink.className).toBe("nav-link selected-nav");
    expect(documentaryLink.className).toBe("nav-link null");
    expect(horrorLink.className).toBe("nav-link null");
    expect(crimeLink.className).toBe("nav-link null");
    // on click of link, all other links will be null and selected selected-nav
    fireEvent.click(documentaryLink);
    expect(documentaryLink.className).toBe("nav-link selected-nav");
    expect(allLink.className).toBe("nav-link null");
    expect(horrorLink.className).toBe("nav-link null");
    expect(crimeLink.className).toBe("nav-link null");
    fireEvent.click(horrorLink);
    expect(horrorLink.className).toBe("nav-link selected-nav");
    expect(documentaryLink.className).toBe("nav-link null");
    expect(allLink.className).toBe("nav-link null");
    expect(crimeLink.className).toBe("nav-link null");
    fireEvent.click(crimeLink);
    expect(horrorLink.className).toBe("nav-link null");
    expect(documentaryLink.className).toBe("nav-link null");
    expect(allLink.className).toBe("nav-link null");
    expect(crimeLink.className).toBe("nav-link selected-nav");
    fireEvent.click(allLink);
    expect(horrorLink.className).toBe("nav-link null");
    expect(documentaryLink.className).toBe("nav-link null");
    expect(allLink.className).toBe("nav-link selected-nav");
    expect(crimeLink.className).toBe("nav-link null");
  });
  test("Dropdown menu initial render", () => {
    const { getByText } = component;
    const dropDownToggle = getByText("SORT BY");
    fireEvent.click(dropDownToggle);
    getByText("Release Date Ascending");
    getByText("Release Date Descending");
    getByText("Highest Rated");
    getByText("Clear Sort");
  });
  test("Dropdown menu click and select Release Date Ascending", () => {
    const { getByText, getByTestId } = component;
    const dropDownToggle = getByTestId("dropdown-toggle");
    fireEvent.click(dropDownToggle);
    const releaseDateAsc = getByText("Release Date Ascending");
    fireEvent.click(releaseDateAsc);
    expect(dropDownToggle.textContent).toContain("RELEASE DATE ASCENDING");
  });
  test("Dropdown menu click and select Release Date Descending", () => {
    const { getByText, getByTestId } = component;
    const dropDownToggle = getByTestId("dropdown-toggle");
    fireEvent.click(dropDownToggle);
    const releaseDateDesc = getByText("Release Date Descending");
    fireEvent.click(releaseDateDesc);
    expect(dropDownToggle.textContent).toContain("RELEASE DATE DESCENDING");
  });
  test("Dropdown menu click and select Highest Rated", () => {
    const { getByText, getByTestId } = component;
    const dropDownToggle = getByTestId("dropdown-toggle");
    fireEvent.click(dropDownToggle);
    const highestRated = getByText("Highest Rated");
    fireEvent.click(highestRated);
    expect(dropDownToggle.textContent).toContain("HIGHEST RATED");
  });
});
