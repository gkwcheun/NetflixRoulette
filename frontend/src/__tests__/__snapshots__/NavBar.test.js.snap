// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Bar Testing Nav Bar Snapshot 1`] = `
<div
  className="nav-bar-container"
>
  <nav
    className="nav-bar"
  >
    <a
      className="nav-link selected-nav"
      href="/search"
      name="all"
      onClick={[Function]}
    >
      ALL
    </a>
    <a
      className="nav-link null"
      href="/search?genre=documentary"
      name="documentary"
      onClick={[Function]}
    >
      DOCUMENTARY
    </a>
    <a
      className="nav-link null"
      href="/search?genre=comedy"
      name="comedy"
      onClick={[Function]}
    >
      COMEDY
    </a>
    <a
      className="nav-link null"
      href="/search?genre=horror"
      name="horror"
      onClick={[Function]}
    >
      HORROR
    </a>
    <a
      className="nav-link null"
      href="/search?genre=crime"
      name="crime"
      onClick={[Function]}
    >
      CRIME
    </a>
  </nav>
  <div
    className="sort-nav"
  >
    <div
      className="dropdown"
    >
      <button
        aria-expanded={false}
        className="sortby-toggle dropdown-toggle btn btn-dark"
        data-testid="dropdown-toggle"
        disabled={false}
        id="dropdown-button-dark-example2"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sort-by-label"
        >
          SORT BY
        </span>
        RELEASE DATE
      </button>
    </div>
  </div>
</div>
`;
